File: 0053. Maximum Subarray/53_naive.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans, tmp_sum, all_neg = 0, 0, True
        for n in nums:
            if n >= 0:
                all_neg = False
                tmp_sum += n

            else:
                if tmp_sum >= -n:
                    tmp_sum += n

                else:
                    tmp_sum = 0

            ans = max(ans, tmp_sum)

        return ans if not all_neg else max(nums)

File: 0053. Maximum Subarray/53_naive.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans = 0, sub_sum = 0, max_n = nums[0];
        bool all_neg = true;
        for(auto const &n: nums){
            if(n>=0){
                all_neg = false;
                sub_sum += n;
            }
            else{
                if(sub_sum >= -n)
                    sub_sum += n;
                else
                    sub_sum = 0;
            }
            ans = max(ans, sub_sum);
            max_n = max(max_n, n);

        }
        return !all_neg ? ans : max_n;
    }
};
